File: Day74/Part6/replay_compression.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

// DP for min length
// dp[i][j] = min length of s[i..j]
int dp[105][105];
string s;

int solve(int i, int j) {
    if (i > j) return 0;
    if (i == j) return 1;
    if (dp[i][j] != -1) return dp[i][j];

    int min_len = j - i + 1;

    // Split
    for (int k = i; k < j; ++k) {
        min_len = min(min_len, solve(i, k) + solve(k + 1, j));
    }

    // Compress
    // If s[i..j] is a repetition of some prefix of length L
    int len = j - i + 1;
    for (int L = 1; L <= len / 2; ++L) {
        if (len % L == 0) {
            bool ok = true;
            for (int k = 0; k < len; ++k) {
                if (s[i + k] != s[i + (k % L)]) {
                    ok = false;
                    break;
                }
            }
            if (ok) {
                min_len = min(min_len, solve(i, i + L - 1));
            }
        }
    }

    return dp[i][j] = min_len;
}

int main() {
    if (cin >> s) {
        int n = s.length();
        for(int i=0; i<n; ++i)
            for(int j=0; j<n; ++j)
                dp[i][j] = -1;
                
        int min_len = solve(0, n - 1);
        cout << (n - min_len) << endl;
    }
    return 0;
}

File: Day74/Part6/replay_compression.py
import sys

# Increase recursion depth
sys.setrecursionlimit(2000)

memo = {}

def solve_dp(s):
    if not s:
        return 0
    if len(s) == 1:
        return 1
    if s in memo:
        return memo[s]
    
    n = len(s)
    min_len = n
    
    # Split
    for k in range(1, n):
        min_len = min(min_len, solve_dp(s[:k]) + solve_dp(s[k:]))
        
    # Compress
    for L in range(1, n // 2 + 1):
        if n % L == 0:
            pattern = s[:L]
            if pattern * (n // L) == s:
                min_len = min(min_len, solve_dp(pattern))
                
    memo[s] = min_len
    return min_len

def solve():
    try:
        line = sys.stdin.read().split()
        if not line:
            return
        s = line[0]
        
        global memo
        memo = {}
        
        final_len = solve_dp(s)
        print(len(s) - final_len)
            
    except Exception:
        pass

if __name__ == "__main__":
    solve()
